chrome/content/zotero/components/tagSelector/tagSelectorList.jsx

```diff
@@ -214,6 +223,91 @@ class TagList extends React.PureComponent {
+
+	shiftFocus(e) {
+		if (!document.activeElement.classList.contains("tag-selector-item")) return;
+		if (!["ArrowRight", "ArrowLeft"].includes(e.key)) return;
+		// Handle arrow navigation
+		let nextTag = (node) => {
+			if (e.key == "ArrowRight") return node.nextElementSibling;
+			return node.previousElementSibling;
+		};
+		let nextOne = nextTag(document.activeElement);
+		// Skip disabled tags
+		while (nextOne && nextOne.classList.contains("disabled")) {
+			nextOne = nextTag(nextOne);
+		}
+		if (nextOne) {
+			nextOne.focus();
+		}
+	}
+
+	// Try to refocus a focused tag that was removed due to windowing
+	refocusTag() {
+		let tagsList = document.querySelector('.tag-selector-list');
+		let tagsNodes = [...tagsList.querySelectorAll(".tag-selector-item")];
+		let tagToFocus = this.props.tags[this.focusedTagIndex];
+		let nodeToFocus = tagsNodes.find(node => node.textContent == tagToFocus.name);
+		if (nodeToFocus) {
+			nodeToFocus.focus();
+		}
+	}
+
+	waitForSectionRender() {
+		return new Promise((resolve, _) => {
+			this.resolveTagsRenderedPromise = resolve;
+		});
+	}
+
+	handleSectionRendered = ({ indices }) => {
+		let tagsList = document.querySelector('.tag-selector-list');
+		// <Collection> sets role="grid" which is not semantically correct
+		tagsList.setAttribute("role", "group");
+
+		if (this.focusedTagIndex === null) return;
+		// If the focused tag does not changed, the scrollToCell won't change
+		// either, so the <Collection> won't scroll to the desired tag if we don't reset it.
+		// E.g. second arrowLeft keypress when first tag is focused won't scroll to it.
+		if (this.lastFocusedTagIndex === this.focusedTagIndex) {
+			this.setState({ scrollToCell: null });
+		}
+		// Check if the tag that is supposed to be focused is within the rendered tags range.
+		// If it is, make sure it is focused. If it is not - focus the tags list.
+		if (indices.includes(this.focusedTagIndex)) {
+			this.refocusTag();
+			if (this.resolveTagsRenderedPromise) {
+				this.resolveTagsRenderedPromise();
+			}
+		}
+		else {
+			tagsList.focus();
+		}
+	};
+
+	handleBlur = (event) => {
+		// If the focus leaves the tags list, clear the last focused tag index
+		let tagsList = document.querySelector('.tag-selector-list');
+		if (!tagsList.contains(event.relatedTarget)) {
+			this.focusedTagIndex = null;
+			this.lastFocusedTagIndex = null;
+		}
+	};
+
+	handleKeyDown(e) {
+		if (!["ArrowRight", "ArrowLeft"].includes(e.key)) return;
+		// If the windowing kicks in, the node of the initially-focused tag may not
+		// exist, so first we may need to scroll to it.
+		if (!document.activeElement.classList.contains("tag-selector-item")) {
+			this.setState({ scrollToCell: this.focusedTagIndex }, async () => {
+				// Even though the state was updated, the tags are not rendered yet.
+				// So we have to wait for handleSectionRendered to run before proceeding.
+				await this.waitForSectionRender();
+				this.shiftFocus(e);
+			});
+			return;
+		}
```

Comment:
So ideally we simplify this further and inline `this.shiftFocus(e)`. You can do this by replacing a callback with a promise like

```js
await new Promise(this.setState({ scrollToCell: this.focusedTagIndex }, () => resolve));
await this.waitForSectionRender();
```

Reply:
Right, no need for a callback if it's not necessary. In fact, I think we might not need to `await` the first promise at all, since we'll be `await`-ing for the `handleSectionRendered` which happens strictly after `setState`. So what if we do just :
```
if (!document.activeElement.classList.contains("tag-selector-item")) {
	this.setState({ scrollToCell: this.focusedTagIndex });
	await this.waitForSectionRender();
	this.shiftFocus(e);
	return;
}
```
I pushed that little tweak and everything still seems to come together with the `handleKeyDown` being way more readable than before

Reply:
Sure, but just get rid of

```js
this.shiftFocus(e);
return;
```

and inline the call to the function (i.e. don't have `this.shiftFocus()` as a separate function) at the end.

Reply:
Oh, because the `shiftFocus` in now only called in `handleKeyDown`, so there's no reason to keep it as a separate function? I moved the rest of focus-shifting logic back into `handleKeyDown` so everything is in one place. Is that what you had in mind?